ingress/controllers/nginx/controller.go

```diff
@@ -217,6 +257,75 @@ func (lbc *loadBalancerController) getUDPConfigMap(ns, name string) (*api.Config
 	return lbc.client.ConfigMaps(ns).Get(name)
 }
 
+func (lbc *loadBalancerController) updateEpNamedPorts(key string) {
+	if !lbc.controllersInSync() {
+		lbc.svcEpQueue.requeue(key, fmt.Errorf("deferring sync till endpoints controller has synced"))
+		return
+	}
+
+	svcObj, svcExists, err := lbc.svcLister.GetByKey(key)
+	if err != nil {
+		glog.Warningf("error getting service %v: %v", key, err)
+		return
+	}
+
+	if !svcExists {
+		glog.Warningf("service %v not found", key)
+		return
+	}
+
+	svc := svcObj.(*api.Service)
+	if svc.Spec.Selector == nil {
+		return
+	}
+
+	pods, err := lbc.client.Pods(svc.Namespace).List(api.ListOptions{
+		LabelSelector: labels.Set(svc.Spec.Selector).AsSelector(),
+	})
+	if err != nil {
+		glog.Errorf("error searching service pods %q: %v", key, err)
+		return
+	}
+
+	namedPorts := map[string]string{}
+
+	for i := range pods.Items {
+		pod := &pods.Items[i]
+
+		for i := range svc.Spec.Ports {
+			servicePort := &svc.Spec.Ports[i]
+
+			_, err := strconv.Atoi(servicePort.TargetPort.StrVal)
+			if err != nil {
+				portNum, err := podutil.FindPort(pod, servicePort)
+				if err != nil {
+					glog.V(4).Infof("Failed to find port for service %s/%s: %v", svc.Namespace, svc.Name, err)
+					continue
+				}
+
+				if servicePort.TargetPort.StrVal == "" {
+					continue
+				}
+
+				namedPorts[servicePort.TargetPort.StrVal] = fmt.Sprintf("%v", portNum)
+			}
+		}
+	}
+
+	if !reflect.DeepEqual(svc.ObjectMeta.Annotations, namedPorts) {
+		data, _ := json.Marshal(namedPorts)
+		if svc.ObjectMeta.Annotations == nil {
+			svc.ObjectMeta.Annotations = map[string]string{}
+		}
+		svc.ObjectMeta.Annotations[namedPortAnnotation] = string(data)
+		glog.Infof("updating service %v with new named port mappings", svc.Name)
+		_, err := lbc.client.Services(svc.Namespace).Update(svc)
```

Comment:
What happens if this update fails? (eg: resource version conflict). And if it passes, will we actuallly just come back in this function and relist pods again?